app/helpers/blacklight_helper.rb

```diff
@@ -31,6 +31,33 @@ def only_advanced_facets(solr_parameters, user_parameters)
     solr_parameters['facet.pivot'] = []
   end
 
+  def course_reserve_filters(solr_parameters, user_parameters)
+    return unless user_parameters[:f]
+    instructor = Array(user_parameters[:f][:instructor_name]).first
+    course = Array(user_parameters[:f][:course]).first
+    department = Array(user_parameters[:f][:department]).first
+    return unless instructor || course || department
+    courses = CourseReserveRepository.all.query(instructor: instructor, course_with_id: course, department_with_identifier: department)
+    index_course_reserves(courses, user_parameters)
+    solr_parameters[:fq] ||= []
+    solr_parameters[:fq].reject! do |x|
+      %w(course instructor_name department).include?(x.split('=')[1].split('}')[0])
+    end
+    solr_parameters[:fq] << "{!join from=bib_ids_s to=id fromIndex=#{ReserveIndexer.core}}#{courses.solr_query}"
+  end
+
+  def index_course_reserves(courses, user_parameters)
+    return unless first_reserve_request(user_parameters)
+    ReserveIndexer.connection.delete_by_query(courses.solr_query)
+    ReserveIndexer.index!(courses)
+  end
+
+  def first_reserve_request(user_parameters)
+    return if user_parameters[:page]
+    return if user_parameters[:f] && !user_parameters[:f].except(:instructor_name, :department, :course).keys.empty?
```

Comment:
Performance is good enough that we probably can get away with indexing on every query.